cmd/createcert/main.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2022-2023 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+// createcert is a simple standalone Go program to generate certificates for
+// use by the conformance tests. It generates a CA, and then generates a
+// client cert, two server certs (for connect and gRPC), and a client/server
+// cert for envoy. The certificates expire after 10 years, and once updated
+// any downstream projects will need to incorporate the new certs to run the
+// conformance tests.
+//
+// It borrows heavily from https://mkcert.dev/ but tweaks the certificates
+// to the types needed for the conformance suite.
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/pem"
+	"fmt"
+	"log"
+	"math/big"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+type CertType int
+
+const (
+	Client CertType = 1 << iota
+	Server
+)
+
+const certDir = "cert"
+
+func main() {
+	// Generate certificates that last 10 years
+	notAfter := time.Now().AddDate(10, 0, 0)
+	var certgen generator
+	if err := certgen.createCA("ConformanceCA", notAfter); err != nil {
+		log.Fatalf("failed to create CA certificate: %v", err)
+	}
+	if err := certgen.createCert("client", Client, notAfter); err != nil {
+		log.Fatalf("failed to create client certificate: %v", err)
+	}
+	if err := certgen.createCert("envoy", Client|Server, notAfter); err != nil {
```

Comment:
Envoy appears to use in both Upstream/Downstream, so I generated a Client/Server cert while the others are either client or server.